### PR TITLE
Rapid Piping Fabrication Device

### PR DESCRIPTION
**Pull request by Geevies**
* Implements a Piping Rapid Fabrication Device, otherwise known as the RFD-P.

They get added to atmos tech lockers. Can't really think of a better place to slap it in. Doesn't come with extra ammo.

### CODE REVIEW

**Comment by Geevies** [comment]
!review

**Comment by flimango** [comment]
Not really sure where else to put this other than DMing you. Any chance we could add one to the CE locker for those of us who are atmospherics focused CEs? Don’t want to steal one from my beloved atmos techs every round.

**Comment by Geevies** [comment]
Can do.

File: code/game/objects/items/weapons/RFD.dm [hunk 1]
```diff
@@ -413,3 +413,125 @@ RFD Mining-Class
...
+	name = "\improper Rapid-Fabrication-Device P-Class"
+	desc = "A RFD, modified to construct pipes and piping accessories."
+	modes = list("Standard Pipes", "Supply Pipes", "Scrubber Pipes", "Devices")
+	var/selected_mode = "Standard Pipes"
+	var/pipe_examine = "Pipe" // used in the examine proc to see what you're putting down at a glance
+	var/selected_pipe = 0 // default is standard pipe, used for the new pipe creation
+	var/build_cost = 1 // this RFD only uses 1 unit of power per pipe, but can be modified if need be in future
+	var/build_delay = 10
+
+	var/list/standard_pipes = list("Pipe" = 0,
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/RFD.dm]: You might wanna comment what the associated numbers are in these lists.

File: code/game/objects/items/weapons/RFD.dm [hunk 2]
```diff
@@ -413,3 +413,125 @@ RFD Mining-Class
...
+			pipe_selection = devices
+	pipe_examine = input(user, "Choose the pipe you want to deploy.", "Pipe Selection") in pipe_selection
+	selected_pipe = pipe_selection[pipe_examine]
+
+/obj/item/rfd/piping/AltClick(mob/user)
+	selected_mode = input(user, "Choose the category you want to change to.", "Pipe Categories") in modes
+	switch(selected_mode)
+		if("Standard Pipes")
+			pipe_examine = "Pipe"
+			selected_pipe = 0
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/RFD.dm]: I don't blame you for not doing it now, but, these pipes should REALLY be defines,

File: code/game/objects/items/weapons/RFD.dm [hunk 3]
```diff
@@ -413,3 +413,125 @@ RFD Mining-Class
...
+/obj/item/rfd/piping/examine(mob/user)
+	. = ..()
+	to_chat(user, FONT_SMALL(SPAN_NOTICE("Change pipe category by Alt clicking, change pipe selection by using in-hand.")))
+	to_chat(user, SPAN_NOTICE("Selected pipe category: <b>[selected_mode]</b>"))
+	to_chat(user, SPAN_NOTICE("Selected pipe: <b>[pipe_examine]</b>"))
+
+/obj/item/rfd/piping/afterattack(atom/A, mob/user, proximity)
+	if(!proximity)
+		return
+	if(istype(get_area(A), /area/shuttle) || istype(get_area(A), /turf/space/transit))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/RFD.dm]: What if it's a space turf?

**Comment by NonQueueingMatt** [comment]
I don't particularly like the input() way of doing this, but, better than nothing I guess.

File: code/game/objects/items/weapons/RFD.dm [hunk 4]
```diff
@@ -413,3 +413,139 @@ RFD Mining-Class
...
+		return FALSE
+
+	new /obj/item/pipe(T, selected_pipe, NORTH)
+
+	playsound(get_turf(src), 'sound/items/Deconstruct.ogg', 50, TRUE)
+	return TRUE
+
+/obj/item/rfd/piping/attack_self(mob/user)
+	playsound(get_turf(src), 'sound/effects/pop.ogg', 50, FALSE)
+	var/list/pipe_selection
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/RFD.dm]:
```suggestion
	var/list/pipe_selection = list()
```

File: code/game/objects/items/weapons/RFD.dm [hunk 5]
```diff
@@ -413,3 +413,139 @@ RFD Mining-Class
...
+							"Scrubber" = 10,
+							"Gas Pump" = 9,
+							"Pressure Regulator" = 15,
+							"High Power Gas Pump" = 16,
+							"Gas Filter" = 13,
+							"Omni Gas Filter" = 27)
+
+/obj/item/rfd/piping/examine(mob/user)
+	. = ..()
+	to_chat(user, FONT_SMALL(SPAN_NOTICE("Change pipe category by Alt clicking, change pipe selection by using in-hand.")))
```
Comment (NonQueueingMatt) [inline code/game/objects/items/weapons/RFD.dm]: This should probably be at the bottom to look cleaner.

File: html/changelogs/geeves-rfd_pp.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+author: Geeves
+
+delete-after: True
+
+changes: 
+  - rscadd: "Implements a Piping Rapid Fabrication Device, otherwise known as the RFD-P."
```
Comment (NonQueueingMatt) [inline html/changelogs/geeves-rfd_pp.yml]: Perhaps add what altclicking does?

**Pull request merged**